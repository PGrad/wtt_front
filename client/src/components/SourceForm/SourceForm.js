/* eslint-disable react-hooks/rules-of-hooks */
import React from 'react';
import { useAuth0 } from '@auth0/auth0-react';
import useAuthUtils from '@/components/Auth/useAuthUtils';
import { useForm, useFormContext, FormProvider } from 'react-hook-form';
import {
  // useSourcesQuery,
  useSourcesMutation,
  useCreateSourcesMutation,
} from '@/api/queries';
import { getData } from '@/api/apiUtils';
import { useQuery } from 'react-query';
// import { FormErrorMessage } from '@/components/Form';

import { NEW_SOURCE_FIELDS, CROSSWALK_FIELDS } from './sourceArrays';
import './SourceForm.scss';

function eliminateSameKeyValuePairs(obj1, obj2, idSourceName) {
  const commonKeys = Object.keys(obj1).filter((key) => key in obj2);
  const newObj = commonKeys.reduce((acc, key) => {
    if (obj1[key] !== obj2[key]) {
      acc[key] = obj1[key];
    }
    return acc;
  }, {});
  return Object.keys(newObj).length > 0 ? { idSourceName, ...newObj } : null;
}

function formatErrors(errors) {
  return Object.entries(errors).reduce((acc, [key, value]) => {
    return `${acc} \n${key}: ${value.message || value.type}, \n`;
  }, '');
}

export function SourceForm({ setOpenEdit, setSource, source, setMessage }) {
  const [error, setError] = React.useState('');

  // Going back to vanilla react-query for now as the documentation is good
  // TODO graphql
  const { data: savedSource } = useQuery(
    ['sources', { idSourceName: source?.idSourceName }],
    getData,
    { enabled: source?.idSourceName !== undefined },
  );

  const { isAuthenticated } = useAuth0();
  const { loginToCurrentPage } = useAuthUtils();
  if (!isAuthenticated) loginToCurrentPage();

  const mutateSources = useSourcesMutation();
  const createSources = useCreateSourcesMutation();

  const defaultValues = {
    source: source?.source || savedSource?.source || {},
    crosswalk: source?.crosswalk || savedSource?.crosswalk || {},
  };

  const methods = useForm({
    mode: 'all',
    defaultValues,
    values: savedSource,
  });
  const {
    formState: { errors },
    watch,
    handleSubmit,
  } = methods;
  watch('formats', 'csv');
  watch('isoAlpha3', savedSource?.source?.isoAlpha3 ?? source?.isoAlpha3);
  watch('idSourceName', '');
  const [saveButton, setSaveButton] = React.useState('Save');
  const [disabled, setDisabled] = React.useState(false);

  const onSubmit = async (data) => {
    setSaveButton('Saving...');
    const exists =
      savedSource?.source?.idSourceName === data?.source?.idSourceName;
    const idSourceName = data?.source?.idSourceName;
    if (exists) {
      const newSource = eliminateSameKeyValuePairs(
        data?.source,
        savedSource?.source,
        idSourceName,
      );
      const newCrosswalk = eliminateSameKeyValuePairs(
        data?.crosswalk,
        savedSource?.crosswalk,
        idSourceName,
      );
      const payload = {
        source: newSource,
        crosswalk: newCrosswalk,
      };

      await mutateSources.mutateAsync(payload);
      setMessage(`Updated ${idSourceName}! Thanks for the update!`);
    } else {
      data.crosswalk['idSourceName'] = data.source['idSourceName'];
      await createSources.mutateAsync(data);
      setMessage(`Saved ${idSourceName}! Thanks for the submission!`);
    }
    setDisabled(false);
    setSaveButton('Save');
    setSource(null);

    setOpenEdit(false);
  };

  const onError = (err) => {
    console.error('Source Form ERROR: ', errors);
    let errorStringSource = errors?.source ? formatErrors(errors.source) : '';
    let errorStringCrosswalk = errors?.crosswalk
      ? formatErrors(errors.crosswalk)
      : '';

    setError(`Source ERRORs: ${errorStringSource}\n\n
      Crosswalk ERRORs: ${errorStringCrosswalk}\n\n`);
  };

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit, onError)}>
        <div className="sourceform">
          <SubmitButtons
            setOpenEdit={setOpenEdit}
            saveButton={saveButton}
            disabled={disabled}
            setSource={setSource}
          />

          {error && (
            <div className="sourceform__alert">
              <h4>{error}</h4>
            </div>
          )}
          <div className="sourceform__section">
            <FormFields fields={NEW_SOURCE_FIELDS} source={source} />
            <FormFields fields={CROSSWALK_FIELDS} crosswalk={source} />
          </div>
        </div>
      </form>
    </FormProvider>
  );
}

function SubmitButtons({ setOpenEdit, setSource, saveButton, disabled }) {
  const handleCancel = () => {
    setOpenEdit(false);
    setSource(null);
  };
  return (
    <div className="sourceform__submit">
      <button type="button" onClick={handleCancel}>
        Cancel
      </button>
      <button type="submit" disabled={disabled}>
        {saveButton}
      </button>
    </div>
  );
}

function FormFields({ fields }) {
  const {
    register,
    formState: { errors },
  } = useFormContext();
  return (
    <div className="sourceform__section-item">
      <div className="sourceform__section-item-header">
        <h2>{fields.header}</h2>
        <h5>{fields.summary}</h5>
        <p>{fields.directions}</p>
      </div>

      {fields?.inputs.map((input) => {
        const { label, name, type, rules } = input;

        // TODO use plugin for dynamic class names
        const fieldClass = errors[name]
          ? `sourceform__field-${type} sourceform__field-error`
          : `sourceform__field-${type}`;

        switch (type) {
          case 'select': {
            return (
              <FormRow label={label} name={name} key={name}>
                <select {...register(name, rules)} className={fieldClass}>
                  {input?.options.map((option) => (
                    <option key={option?.label} value={option?.value}>
                      {option.label}
                    </option>
                  ))}
                </select>
              </FormRow>
            );
          }

          case 'checkbox': {
            return (
              <FormRow label={label} name={name}>
                <input
                  type="checkbox"
                  {...register(name, rules)}
                  className={fieldClass}
                />
              </FormRow>
            );
          }

          default: {
            return (
              <FormRow label={label} name={name}>
                <input
                  type={type}
                  {...register(name, rules)}
                  className={fieldClass}
                  placeholder={label}
                />
              </FormRow>
            );
          }
        }
      })}
    </div>
  );
}

const errorByType = (label, type) =>
  ({
    required: `${label} is required.`,
    minLength: `${label} is too short.`,
    maxLength: `${label} is too long.`,
    pattern: `${label} is not formatted correctly.`,
    default: `${label} has an error.`,
  }[type ?? 'default']);

function FormRow({ children, label, name }) {
  const {
    formState: { errors },
  } = useFormContext();

  // TODO fix errors. Why are errors not showing up?
  const fieldClassErrors = 'sourceform__field-error';

  return (
    <li className="sourceform__form-row" key={name}>
      <label>{label}</label>
      {children}
      {errors && errors[name] && (
        <span className={fieldClassErrors}>
          {errorByType(label, errors[name].type)}
        </span>
      )}
    </li>
  );
}
