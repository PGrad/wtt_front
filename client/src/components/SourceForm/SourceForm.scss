.sourceform {
    $self: &;
    display: flex;
    justify-content: flex-start;
    flex-direction: column;
    list-style-type: none;
    padding: 0;
    border-radius: 3px;

    input[type=checkbox] { flex: unset; }

    &__section {
        display: flex;
        justify-content: space-between;
        flex-direction: row;
    }
    &__section-item {
        flex-grow: 1;
        flex-basis: 50%;
        margin: 10px;
        padding: 15px;
        border-radius: 5px;
        background-color: lightgray;
    }
    &__section-item-header {
        background-color: darkgray;
        padding: 15px;
        text-align: left;
        margin-bottom: 15px;
        border-radius: 5px;
    }

    &__field-error {
        color: red;
    }

    &__field-error > input,
    &__field-error > select,
    &__field-error > checkbox {
        background: rgba(0,0,0,0.5);
    }

    &__form-row {
        display: flex;
        justify-content: flex-start;
        flex-direction: row;
        margin: 5px 0;
        flex: 4;
    }
    &__form-row > label {
        padding: 0.1em 0;
        flex: 2;
        text-align: start;
    }
    &__form-row > checkbox {
        justify-items: flex-start;
        flex-direction: row;
    }
    &__form-row > select {
        flex: 4;
        width: 100%;
        padding: 0.2em;
        height: 100%;
    }
    &__form-row > input,
    &__form-row > checkbox,
    &__form-row > button {
        padding: .2em;
        flex: 4;
    }

    &__submit {
        text-align: right;
        margin: 0 10px 0 5px;
    }
    &__submit > button {
        border: 0;
        width: fit-content;
        border-radius: 5px;
        font-size: large;
        padding: 0 5px;
        margin: 0 0 0 5px;
    }
    &__submit > button[type="button"] {
        background: gray;
    }
    &__submit > button[type="submit"] {
        background: green;
    }
    &__submit > button:hover {
        background: lightgray;        
    }

    @media screen and (max-width: 768px) {
        display: flex;
        justify-content: flex-start;
        flex-direction: column-reverse;
        margin: 5px 0;

        &__section {
            flex-direction: column;
        }
        &__section-item {
            margin: 5px;
            padding: 5px;
        }
        &__form-row {
            flex-direction: column;
        }
        &__form-row > label {
            flex: 1;
        }
        &__form-row > select,
        &__form-row > input,
        &__form-row > checkbox,
        &__form-row > button {
            flex: 1;
        }
        &__submit {
            text-align: center;
        }
    }
        
    

}
