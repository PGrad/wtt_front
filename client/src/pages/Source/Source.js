import React, { useState, useEffect } from 'react';
// import SourceList from './SourceList';
import SourceHeader from './SourceHeader';
import LicenseFooter from '@/pages/License/LicenseFooter';
import { CSVImporter } from '@/components/CSVImporter/CSVImporter';
import { SourceForm } from '@/components/SourceForm/SourceForm';
import { ListGrid } from '@/components/ListGrid/ListGrid';
import { useSourcesQuery } from '@/api/queries';
import { searchArray } from '@/components/SearchBar/SearchBar';
import './Source.scss';
// import './SourceList.scss';

function createColumnHeaders(source) {
  if (!source) return [];
  return [
    {
      columnHeader: 'idSourceName',
      label: '',
    },
    {
      columnHeader: 'broken',
      label: 'Up',
    },
    {
      columnHeader: 'city',
      label: 'Location',
    },
    {
      columnHeader: 'isoAlpha3',
      label: 'Country',
    },
    {
      columnHeader: 'state',
      label: 'State',
    },

    {
      columnHeader: 'email',
      label: 'Email',
    },
    {
      columnHeader: 'contact',
      label: 'Contact',
    },
    {
      columnHeader: 'phone',
      label: 'Phone',
    },

    {
      columnHeader: 'notes',
      label: 'Notes',
    },
    // TODO turn this on for San Francisco's Friends of the Urban Forest
    // {
    //   columnHeader: 'maintenance',
    //   label: 'Maintenance',
    // },
  ];
}

export default function Source() {
  const [search, setSearch] = useState('');
  const [license, setLicense] = useState('');
  const [importCSV, setImportCSV] = useState(false);
  const [openEdit, setOpenEdit] = useState(false);
  const [source, setList] = useState({});
  const [message, setMessage] = useState('');
  const { data: sources } = useSourcesQuery();
  console.log(sources, 'sources');
  const columns = createColumnHeaders(sources);
  const sourceFiltered = search && searchArray(sources, search);

  const handleNewEditSource = (e, item) => {
    e.preventDefault();
    setOpenEdit(!openEdit);
    setList(item);
  };

  useEffect(() => {
    if (message) {
      setTimeout(() => {
        setMessage('');
      }, 5000);
    }
  }, [message]);

  return (
    <div className="source">
      <SourceHeader
        search={search}
        setSearch={setSearch}
        importCSV={importCSV}
        setImportCSV={setImportCSV}
        setOpenEdit={setOpenEdit}
        openEdit={openEdit}
        setList={setList}
      />

      {message && (
        <div className="message">
          <h3>{`${source.idSourceName} ${message}`}</h3>
        </div>
      )}

      {openEdit && (
        <div className="sources__openedit">
          <SourceForm
            setOpenEdit={setOpenEdit}
            source={source}
            setList={setList}
            message={message}
            setMessage={setMessage}
          />
        </div>
      )}

      {importCSV && <CSVImporter setImportCSV={setImportCSV} />}

      {!openEdit && sources && (
        <ListGrid
          data={sourceFiltered || sources}
          columns={columns}
          defaultSortIndex={0}
          setList={setList}
          openEdit={openEdit}
          setOpenEdit={setOpenEdit}
          handleNewEditSource={handleNewEditSource}
          listType="source"
        />
      )}

      <LicenseFooter license={license} />
    </div>
  );
}
