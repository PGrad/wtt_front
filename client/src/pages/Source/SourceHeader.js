import React from 'react';
import ListGridHeader from '@/components/ListGridHeader/ListGridHeader';
import { GrayButton } from '@/components/Button/Button';
import CloudUploadIcon from '@mui/icons-material/CloudUpload';
import { LinkIcon } from '@/components/Icons';
import './Source.scss';

export default function SourceHeader({
  search,
  setSearch,
  importCSV,
  setImportCSV,
  openEdit,
  setOpenEdit,
  setList,
}) {
  // TODO turn this back on after MVP release
  const featureFlagCSVImporter = false;

  const handleClick = () => {
    setOpenEdit(!openEdit);
    if (openEdit) setList(null);
  };

  return (
    <ListGridHeader
      search={search}
      setSearch={setSearch}
      title={'Sources'}
      description={'Add, edit, and delete sources'}
      searchLabel={'Search Sources'}
    >
      <div className="listgridheader-content-item">
        <GrayButton
          aria-label="Link City/Source"
          size="large"
          onClick={handleClick}
        >
          <LinkIcon color="gray" fontSize="large" aria-label="Download CSV" />
          Link Source
        </GrayButton>
      </div>

      {featureFlagCSVImporter && (
        <div className="listgridheader-content-item">
          <GrayButton
            aria-label="Import City/Source CSV"
            size="small"
            onClick={() => setImportCSV(!importCSV)}
          >
            <CloudUploadIcon sx={{ marginRight: '7px', fontSize: 24 }} />
            Upload Source
          </GrayButton>
        </div>
      )}
    </ListGridHeader>
  );
}
