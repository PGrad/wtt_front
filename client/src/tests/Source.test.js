import React from 'react';
import { render } from '@testing-library/react';
import Source from '@/pages/Source/Source';
import SourceHeader from '@/pages/Source/SourceHeader';
import { SourceForm } from '@/components/SourceForm/SourceForm';

/**
 * These snapshots are designed to catch _any_ change to UI components, so they
 * are by definition very brittle. If the changes are related to your changes
 * and look expected, update these snapshots with `npm test -- -u`
 */

const mockSetOpenEdit = jest.fn(() => {});
const mockSetImportCSV = jest.fn(() => {});
const mockSetList = jest.fn(() => {});
const mockSetSearch = jest.fn(() => {});
// const mockSetMessage = jest.fn(() => {});

describe('<Source /> spec', () => {
  it('renders Source Header correctly', () => {
    const sourceHeader = render(
      <SourceHeader
        search={'test'}
        setSearch={mockSetSearch}
        setImportCSV={mockSetImportCSV}
        setOpenEdit={mockSetOpenEdit}
        setList={mockSetList}
        openEdit={false}
        importCSV={false}
      />,
    );
    expect(sourceHeader).toMatchSnapshot();
  });

  // TODO Need Auth provider to render Source Form
  // it('renders Source Form correctly with null source', () => {
  //   const sourceForm = render(
  //     <SourceForm
  //       setOpenEdit={mockSetOpenEdit}
  //       source={null}
  //       setList={mockSetList}
  //       message={'test message'}
  //       setMessage={mockSetMessage}
  //     />,
  //   );
  //   expect(sourceForm).toMatchSnapshot();
  // });
});
