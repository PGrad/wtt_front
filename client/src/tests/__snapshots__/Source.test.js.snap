// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Source /> spec renders Source Header correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  color: #00000050;
  font-size: 24px;
  margin-left: 10px;
  margin-right: 10px;
  font-family: Montserrat;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #1976d2;
  background-color: white;
  border-radius: .3vw;
  outline: 1px solid #00000050;
  color: #00000050;
  height: 32px;
  font-size: 12px;
  min-width: 150px;
}

.emotion-1::-moz-focus-inner {
  border-style: none;
}

.emotion-1.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-1 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25, 118, 210, 0.04);
}

@media (hover: none) {
  .emotion-1:hover {
    background-color: transparent;
  }
}

.emotion-1.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 2.1875rem;
  vertical-align: top;
}

.emotion-3 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<body>
    <div>
      <div
        class="listgridheader"
      >
        <div
          class="listgridheader-title"
        >
          <h2>
            Sources
          </h2>
          <h5>
            Add, edit, and delete sources
          </h5>
          <div
            class="listgridheader-search"
          >
            <div
              class="searchbar"
              style="border-radius: .3vw; font-size: large;"
            >
              <label
                for="searchbar"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-0"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </label>
              <input
                class="searchbar__input"
                id="searchbar"
                placeholder="Search Sources"
                style="border-radius: .3vw; width: 100%;"
                value="test"
              />
            </div>
          </div>
        </div>
        <div
          class="listgridheader-content"
        >
          <div
            class="listgridheader-content-item"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium emotion-1"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                aria-label="Download CSV"
                class="MuiSvgIcon-root MuiSvgIcon-colorGray MuiSvgIcon-fontSizeLarge emotion-2"
                data-testid="LinkIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
                />
              </svg>
              Link Source
              <span
                class="MuiTouchRipple-root emotion-3"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": .emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  color: #00000050;
  font-size: 24px;
  margin-left: 10px;
  margin-right: 10px;
  font-family: Montserrat;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #1976d2;
  background-color: white;
  border-radius: .3vw;
  outline: 1px solid #00000050;
  color: #00000050;
  height: 32px;
  font-size: 12px;
  min-width: 150px;
}

.emotion-1::-moz-focus-inner {
  border-style: none;
}

.emotion-1.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-1 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25, 118, 210, 0.04);
}

@media (hover: none) {
  .emotion-1:hover {
    background-color: transparent;
  }
}

.emotion-1.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 2.1875rem;
  vertical-align: top;
}

.emotion-3 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
    <div
      class="listgridheader"
    >
      <div
        class="listgridheader-title"
      >
        <h2>
          Sources
        </h2>
        <h5>
          Add, edit, and delete sources
        </h5>
        <div
          class="listgridheader-search"
        >
          <div
            class="searchbar"
            style="border-radius: .3vw; font-size: large;"
          >
            <label
              for="searchbar"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-0"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </label>
            <input
              class="searchbar__input"
              id="searchbar"
              placeholder="Search Sources"
              style="border-radius: .3vw; width: 100%;"
              value="test"
            />
          </div>
        </div>
      </div>
      <div
        class="listgridheader-content"
      >
        <div
          class="listgridheader-content-item"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium emotion-1"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              aria-label="Download CSV"
              class="MuiSvgIcon-root MuiSvgIcon-colorGray MuiSvgIcon-fontSizeLarge emotion-2"
              data-testid="LinkIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
              />
            </svg>
            Link Source
            <span
              class="MuiTouchRipple-root emotion-3"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
